### ISSUE TITLE
Propagate asynchrony through member access expressions

### ISSUE THREAD

**Issue by AnthonyDGreen** [dotnet/vblang#116]
It's inconvenient to access a member of the result of an Await operation because you have to backtrack (I hate backtracking) and wrap the entire expression in parenthesis. Once I recall having to even write a call like this:

`Dim si = (Await (Await context.GetDocumentAsync()).GetSemanticModalAsync()).GetSymbolInfo(node)`

Visual Basic is nothing if not a constant struggle against the tyranny of parentheses. I think we can apply a similar trick to what we do for method calls and just elide some intermediate parentheses so you could write the above as:

`Dim si = Await context.GetDocumentAsync().GetSemanticModelAsync().GetSymbolInfo(node)`

So basically, if a member access on an awaitable type fails _and_ that member access would succeed on the type of GetAwaiter().GetResult() we insert an implicit Await _but_ the result type of the expression becomes "asynchronous" and must ultimately be awaited. I think that cleans up the code without harming readability, particularly if the -Async naming convention is used.

Of course, for this to work at all for class libraries we'd need to do #115

**Comment by ericmutta** [comment]
@AnthonyDGreen ... because you have to backtrack (I hate backtracking) and wrap the entire expression in parenthesis.

Add me to the team of people who hate backtracking to insert parenthesis on an expression just to access a member of the resulting value!

I can't comment specifically on the `Await` case (this is probably the only new feature of VB that I haven't used in real code, because the compiler is fussy about where you can use that keyword in ways that I never found intuitive...so I ended up doing a small library of [futures and promises](https://en.wikipedia.org/wiki/Futures_and_promises) which uses ordinary objects which I **can** "dot into" to access members without backtracking).

...but there is another scenario where I can comment: LINQ queries. Say you start with this:

```vb
Dim query = From n In Enumerable.Range(1, 10) Where n Mod 2 = 0
```

If you decide you actually want a count of the items rather than a list you have to go back and wrap the whole thing in brackets:

```vb
Dim query = (From n In Enumerable.Range(1, 10) Where n Mod 2 = 0).Count()
```

In a similar way to the `Await` situation it would be handy if we had a member access operator that would "auto-bracket" its left operand. For example if we used the operator you suggest in #165:

```vb
Dim query = From n In Enumerable.Range(1, 10) Where n Mod 2 = 0 -> Count()
```

This would be really handy, since you can switch from listing to counting to anything else just by appending/removing `-> SomeMethod()`. No need to fiddle with brackets!

PS: such a special operator could be made type specific (i.e works only on expressions of IEnumerable/IQueryable ) similar to the way XML literals have `...` and `...<foo>` and `.@foo` member access operators that only work for XElement objects.

PPS: for the `Await` case (which again, I have little experience with) the same `->` operator would allow awaits to be chained:

```vb
Dim si = context.GetDocumentAsync() -> GetSemanticModelAsync() -> GetSymbolInfo(node)
```

...the above intuitively reads as get the document then wait, get the semantic model from that then wait and finally get the symbol info from that and put the value into `si`.

**Comment by bandleader** [comment]
@ericmutta General backtracking prevention now has its own proposal: #154 

```vb
Dim si = Await context.GetDocumentAsync().GetSemanticModelAsync().GetSymbolInfo(node)
```

@AnthonyDGreen While I agree with the premise of the idea, I find the above-proposed code to be unclear -- it still looks like you're awaiting the result of `GetSymbolInfo(node)`.

I would propose instead one of the following:

1. Solve this scenario by piping the async function into Await, as I suggested in #154:
```vb
Dim si = context.GetDocumentAsync()->Await.GetSemanticModelAsync()->Await.GetSymbolInfo(node)
'Or IMHO '..' is more natural:
Dim si = context.GetDocumentAsync()..Await.GetSemanticModelAsync()..Await.GetSymbolInfo(node)
'Maybe even make Await() usable as a virtual function, similar to CType, DirectCast, and If.
'Much more readable IMHO and still understandable:
Dim si = context.GetDocumentAsync()..Await().GetSemanticModelAsync()..Await().GetSymbolInfo(node)
```
2. If you indeed feel that propogation of asynchrony is important, use a new `AwaitAll` keyword that makes each async call in the expression automatically Awaited:
```vb
Dim si = AwaitAll context.GetDocumentAsync().GetSemanticModelAsync().GetSymbolInfo(node)
```
